R/disease_progression.R

```diff
@@ -0,0 +1,64 @@
+#' @title Modelling the progression of the human disease
+#' @description Randomly moves individuals towards the later stages of disease
+#' and updates their infectivity
+#'
+#' @param human the handle for the human individuals
+#' @param from_state the source disease state
+#' @param to_state the destination disease state
+#' @param rate the rate at which to move humans
+#' @param infectivity the handle for the infectivity variable
+#' @param new_infectivity the new infectivity of the progressed individuals
+create_progression_process <- function(
+ human,
+ from_state,
+ to_state,
+ rate,
+ infectivity,
+ new_infectivity
+ ) {
+  function(api) {
+    source_humans <- api$get_state(human, from_state)
+    to_move <- source_humans[bernoulli(length(source_humans), rate)]
```

Comment:
Looking at this, could you use a partial shuffle here more efficiently? As you have it `bernoulli` will do one draw per individual, but you'd draw on the order of `nh * rate` numbers with:

```
nh <- length(source_humans)
sample.int(nh, rbinom(1, nh, rate))
```

not sure if this is an important time sink of course. Similar comments apply wherever else you use bernoulli in order to identify individuals to move

Reply:
Oh this is fantastic, thanks! It _is_ a timesink